camonitor of a string array PV raises a TypeError exception if the array contains one string

Exception:
```
File "...\site-packages\epics\__init__.py", line 157, in camonitor
  thispv.get()
File "...\site-packages\epics\pv.py", line 486, in get
  data = self.get_with_metadata(count=count, as_string=as_string,
File "...\site-packages\epics\pv.py", line 48, in wrapped
  return func(self, *args, **kwargs)
File "...\site-packages\epics\pv.py", line 621, in get_with_metadata
  if count < len(val):
TypeError: len() of unsized object
```

The behavior is a result of interaction between following functions:
* `ca.unwap()` uses `ca.scan_string` to read a string. This function will return the string if retrieved list of strings contains just one string.
* `PV.get_with_metadata(as_numpy=True)` (the default used by `camonitor()`) does following:
  * Reads the string into a variable `val`
  * It tests that `val` supports `__len__`. In case of a string, this passes.
  * Then, it verifies `as_numpy` and `ca.HAS_NUMPY`. In this is satisfied. So, it converts `val` to a numpy object.
  * Then, it tries to run `len(val)`. Yet, `val` is a scalar at this moment (`val.shape` returns `()`). This raises an exception.

Further details:
* This has been observed on pyepics 3.4.3 using for automated tests. Yet, the code involved in this mechanism is the same at the current master branch.
* Standalone recreation of the scenario has not been performed because the PV is a custom record which mimics 
waveform (FTVL=STRING): it provides an array of strings. caget works with no complains.
* Use of `as_numpy=False` with `caget()` will return the first letter of the first string (i.e. `"R"` instead of `"READY"`) for an array containing `["READY"]`.

@newville Thank you for your response. Let me add some details:
* The custom record type provides the same fields as waveform (VAL, NELM, NORD) and it uses almost the same routines `cvt_dbaddr()` and `get_array_info()` as base waveform and EPICS tools caget/caput works with it. 
* NELM is 32, NORD is 1, and the record contains `["ITEM 1"]`. 

I was able to trigger the problem by using a plain waveform (see test `test_pv_unittests.py/test_get_string_waveform` in https://github.com/hanak/pyepics/tree/waveform-string-error). This test fails with:
```
FAILED tests/test_pv_unittests.py::test_get_string_waveform - TypeError: len() of unsized object
```

Can you confirm you see the same?

One I remove the case 1 from the pool of lengths, the test succeeded. Therefore, it seems to me this issue is generic rather than specific to the custom record type of mine. I think that the problem is a result of using `len()` as a tool for detecting an array combined with converting an array of one string to the string. My guess is the solution is to modify use of `len()`. Still, I am not able to tell whether this will not trigger a surprise somewhere else.

Environment:
* OS: Windows 10
* python 3.10.11
* pytest 8.2.2
* softIoc of EPICS base 7.0.6.1

@newville Thank you for your response. Let me add some details:

> The custom record type provides the same fields as waveform (VAL, NELM, NORD) and it uses almost the same routines 
> cvt_dbaddr() and get_array_info() as base waveform and EPICS tools caget/caput works with it.
> NELM is 32, NORD is 1, and the record contains ["ITEM 1"].

Sorry, that is not enough to go on. If you would like us to support using such a record, you will need to post code
to allow someone else to create such a record. 

Is there such a thing as a waveform of strings?   


>  I was able to trigger the problem by using a plain waveform (see test test_pv_unittests.py/test_get_string_waveform in  
> https://github.com/hanak/pyepics/tree/waveform-string-error). This test fails with:
>
> FAILED tests/test_pv_unittests.py::test_get_string_waveform - TypeError: len() of unsized object
> Can you confirm you see the same?

No, I cannot confirm that.  

If you see a problem that you want us to act on, post real code that we can run and reproduce the problem.

--- a/epics/ca.py
+++ b/epics/ca.py
@@ -751,7 +751,6 @@ def _onMonitorEvent(args):
         kwds.update(**_unpack_metadata(ftype=args.type, dbr_value=value[0]))
     except IndexError:
         pass
-
     value = _unpack(args.chid, value, count=args.count, ftype=args.type)
     if callable(args.usr):
         args.usr(value=value, **kwds)
@@ -1260,7 +1259,7 @@ def _unpack(chid, data, count=None, ftype=None, as_numpy=True):
         whether to convert to numpy array.
     """
 
-    def scan_string(data, count):
+    def scan_string(data, count, elem_count):
         """ Scan a string, or an array of strings as a list, depending on content """
         out = []
         for elem in range(min(count, len(data))):
@@ -1268,7 +1267,7 @@ def _unpack(chid, data, count=None, ftype=None, as_numpy=True):
             if '\x00' in this:
                 this = this[:this.index('\x00')]
             out.append(this)
-        if len(out) == 1:
+        if len(out) == 1 and elem_count==1:
             out = out[0]
         return out
 
@@ -1294,7 +1293,7 @@ def _unpack(chid, data, count=None, ftype=None, as_numpy=True):
         if count == 1 and ntype != dbr.STRING:
             return data[0]
         if ntype == dbr.STRING:
-            return scan_string(data, count)
+            return scan_string(data, count, elem_count)
         if count != 1:
             return array_cast(data, count, ntype, use_numpy)
         return data

--- a/tests/Setup/simulator.py
+++ b/tests/Setup/simulator.py
@@ -120,9 +120,9 @@ while True:
 
     t0 = time.time()
     if pause_pv.get() == 1:
-        # pause for up to 120 seconds if pause was selected
+        # pause for up to 300 seconds if pause was selected
         t0 = time.time()
-        while time.time()-t0 < 120:
+        while time.time()-t0 < 300:
             time.sleep(SLEEP_TIME)
             if pause_pv.get() == 0:
                 break

--- a/tests/test_pv_unittests.py
+++ b/tests/test_pv_unittests.py
@@ -306,16 +306,12 @@ def test_string_waveform_lengths():
     readback = spv.get()
     nelm = spv.nelm
     with no_simulator_updates():
-        n = 0
-        for count in (nelm, nelm-2, nelm//3, 11, 8, 5, 3, 2, 1):
-            n += 5*count
-            dat = [f"str_{i+n}" for i in range(count)]
+        for count in (nelm, nelm-5, nelm//3, 11, 5, 1, 3, 7, 2, 1):
+            dat = [f"str_{count}_{i}" for i in range(count)]
             spv.put(dat)
-
             time.sleep(0.1)
             readback = spv.get()
             assert all(readback == dat)
-            #            print(count, dat, readback)
 
 
 def test_subarrays():

--- a/tests/test_pv_unittests.py
+++ b/tests/test_pv_unittests.py
@@ -6,7 +6,7 @@ import time
 import numpy
 import threading
 import pytest
-
+from random import random
 from contextlib import contextmanager
 from epics import PV, get_pv, caput, caget, caget_many, caput_many, ca
 
@@ -267,7 +267,7 @@ def test_put_string_waveform_single_element():
         pv.put(put_value, wait=True)
         time.sleep(0.05)
         get_value = pv.get(use_monitor=False, as_numpy=False)
-        assert put_value[0] == get_value
+        assert put_value == get_value
 
 def test_put_string_waveform_mixed_types():
     write('String Array: put mixed types\n')
@@ -298,6 +298,25 @@ def test_put_string_waveform_zero_length_strings():
         time.sleep(0.05)
         get_value = pv.get(use_monitor=False, as_numpy=False)
         numpy.testing.assert_array_equal(get_value, put_value)
+
+
+def test_string_waveform_lengths():
+    "test string arrays with length=1 and other lengths"
+    spv = get_pv(pvnames.string_arr_pv)
+    readback = spv.get()
+    nelm = spv.nelm
+    with no_simulator_updates():
+        n = 0
+        for count in (nelm, nelm-2, nelm//3, 11, 8, 5, 3, 2, 1):
+            n += 5*count
+            dat = [f"str_{i+n}" for i in range(count)]
+            spv.put(dat)
+
+            time.sleep(0.1)
+            readback = spv.get()
+            assert all(readback == dat)
+            #            print(count, dat, readback)
+
 
 def test_subarrays():
     write("Subarray test:  dynamic length arrays\n")

